### PR TITLE
ISPN-14974 OpenTelemetry Tracing 2.0 - part 1/3

### PR DESCRIPTION
**Pull request by fax4ever**
https://issues.redhat.com/browse/ISPN-14974

Based on #11474.
Base PR for https://github.com/infinispan/infinispan/pull/11308.

We're creating here all the base infrastructure to address the rest of the work.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -661,6 +666,11 @@
             <artifactId>opentelemetry-exporter-otlp-common</artifactId>
             <version>${version.opentelemetry}</version>
          </dependency>
+         <dependency>
+            <groupId>io.opentelemetry</groupId>
+            <artifactId>opentelemetry-exporter-sender-okhttp</artifactId>
```
Comment (pruivo) [inline pom.xml:671]: should be replaced with `io.opentelemetry:opentelemetry-exporter-sender-jdk`

File: client/hotrod-client/pom.xml [hunk 2]
```diff
@@ -308,6 +302,29 @@
          <plugin>
             <groupId>org.apache.maven.plugins</groupId>
             <artifactId>maven-surefire-plugin</artifactId>
+            <executions>
```
Comment (pruivo) [inline client/hotrod-client/pom.xml:305]: why is this required for this module and not the other hot rod client module?

**Review by pruivo — CHANGES_REQUESTED** [review]
just a quick look into pom.xml files

File: server/core/src/main/java/org/infinispan/server/core/telemetry/TelemetryServiceFactory.java [hunk 3]
```diff
@@ -40,8 +53,8 @@ public Object construct(String name) {
          log.telemetryLoaded(openTelemetry);
```
Comment (tristantarrant) [inline server/core/src/main/java/org/infinispan/server/core/telemetry/TelemetryServiceFactory.java:53]: Side note: we should only log when OpenTelemetry is enabled, not disabled. Also it should use the `org.infinispan.SERVER` category.

Comment (fax4ever) [reply server/core/src/main/java/org/infinispan/server/core/telemetry/TelemetryServiceFactory.java:49]: done the change thank you

**Comment by fax4ever** [comment]
Thank you @tristantarrant and @pruivo. For the review

File: server/core/src/main/java/org/infinispan/server/core/telemetry/inmemory/InMemoryTelemetryClient.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+import java.util.List;
+
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.SpanKind;
+import io.opentelemetry.api.trace.StatusCode;
+import io.opentelemetry.api.trace.Tracer;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.sdk.trace.data.SpanData;
+
+public class InMemoryTelemetryClient {
```
Comment (pruivo) [inline server/core/src/main/java/org/infinispan/server/core/telemetry/inmemory/InMemoryTelemetryClient.java:12]: It looks like this is only used for tests. It should be moved to the test sources.

**Review by pruivo — CHANGES_REQUESTED** [review]

File: client/hotrod-client/pom.xml (hunk 2 above)
Comment (fax4ever) [reply client/hotrod-client/pom.xml:305]: The other module does not have the same tests :)

File: server/core/src/main/java/org/infinispan/server/core/telemetry/inmemory/InMemorySpanExporter.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    return new InMemorySpanExporter();
+  }
+
+  /**
+   * Returns a {@code List} of the finished {@code Span}s, represented by {@code SpanData}.
+   *
+   * @return a {@code List} of the finished {@code Span}s.
+   */
+  public List<SpanData> getFinishedSpanItems() {
+    return Collections.unmodifiableList(new ArrayList<>(finishedSpanItems));
```
[the remark this answers is not in the document]
Comment (fax4ever) [reply server/core/src/main/java/org/infinispan/server/core/telemetry/inmemory/InMemorySpanExporter.java:64]: This is not my code, I copied and adapt from the OpenTelemetry client. Anyway I'm applying the suggestions.

**Comment by fax4ever** [comment]
Thank you @pruivo for the further review. I think I've addressed all the ideas here.

**Comment by fax4ever** [comment]
With the last push I dropped the commit `ISPN-14974 Initialize at runtime an OTEL class`. It seems that we don't need at this stage.

**Review by pruivo — APPROVED** [review]

**Comment by pruivo** [comment]
@fax4ever there are a lot of failures in CI!

File: server/hotrod/src/main/java/org/infinispan/server/hotrod/CacheRequestProcessor.java [hunk 6]
```diff
@@ -661,11 +687,18 @@ void putStream(HotRodHeader header, Subject subject, byte[] key, ByteBuf buf, lo
...
-         telemetryService.recordException(span, cause);
+         span.recordException(cause);
       } finally {
          writeException(header, cause);
       }
    }
+
+   private InfinispanSpan requestStart(HotRodHeader header, AdvancedCache<?, ?> cache) {
+      var attributes = new InfinispanSpanAttributes.Builder(SpanCategory.CONTAINER)
+            .withCache(header.cacheName, cache.getCacheConfiguration())
```
Comment (pruivo) [inline server/hotrod/src/main/java/org/infinispan/server/hotrod/CacheRequestProcessor.java:700]: Looks like this line has to be changed to:
```suggestion
            .withCache(header.cacheName, SecurityActions.getCacheConfiguration(cache))
```

Comment (fax4ever) [reply server/hotrod/src/main/java/org/infinispan/server/hotrod/CacheRequestProcessor.java]: applied the change thank you

**Comment by pruivo** [comment]
merged! thanks @fax4ever !

**Comment by fax4ever** [comment]
Thanks to you @pruivo!
